File: graph.cpp
#include <iostream>
using namespace std;

class LinkedListNode{
public:
	int data;
	LinkedListNode* next;
	LinkedListNode* prev;

	LinkedListNode(int data){
		this->data = data;
		this->next = nullptr;
		this->prev = nullptr;
	}
};

class LinkedList {
public :
	LinkedListNode* head;
	LinkedListNode* tail;

	LinkedList(){
		this->head = nullptr;
		this->tail = nullptr;
		//cout << ":)" << endl;
	}

	void insert(int data){
		LinkedListNode* node = new LinkedListNode(data);

		if(this->head == nullptr){
			this->head = node;
			this->tail = node;
		}

		else{
			node->prev = tail;
			node->next = nullptr;
			this->tail->next = node;
			this->tail = node;
		}

	}

	void remove(){
		if(this->head == nullptr){
			return;
		}

		else if(this->head == this->tail){
			LinkedListNode* del = this->head;
			head = nullptr;
			tail = nullptr;
			delete del;
		}
		else{
			LinkedListNode* del = this->head;
			this->head = this->head->next;
			this->head->prev = nullptr;
			delete del;
		}
	}

	void printLinkedList(){
		LinkedListNode* temp = this->head;
		while(temp->next != nullptr){
			cout << temp->data << "->"  ;
			temp = temp->next;
		}
		cout << temp->data << endl;
		temp = temp->next;
		free(temp);	
	}
};

class Graph{
public:
	LinkedList* adjlist;
	int vertices;
	
	Graph(int v){
		vertices = v;
	    adjlist = new LinkedList[v]();
	    //cout << "All Good!!" << endl;
	}

	void addEdge(int src , int dest){
		adjlist[src].insert(dest);
	}

	void printGraph(){
		for(int i = 0 ; i < this->vertices ; i++ ){
			cout << i << "->" ;
			adjlist[i].printLinkedList();
		}
	}

	~Graph(){
		
		delete adjlist;
	}

};

int main(){

    int v = 3;
	Graph* g1 = new Graph(v) ; 

	g1->addEdge(0,1);
	g1->addEdge(0,2);
	g1->addEdge(1,0);
	g1->addEdge(2,1);
	
    g1->printGraph();

    delete g1;
}

File: graph.py
# -*- coding: utf-8 -*-
"""
Created on Sun May  5 22:26:45 2019

@author: HP
"""

from collections import deque


class Graph:

    def __init__(self , vertices):
        self.vertices = vertices
        self.g = {}
        for i in range(0, vertices):
            self.g[i] = []
        
    def addEdge(self , src , dest):
        self.g[src].append(dest)
        
    def printGraph(self):
        print(self.g)
        print(self.g.keys())
        for key,items in self.g.items():
            print(key,end = '')
            for i in items:
                print('->',i,sep = '',end = '')
            print()
                      
            
    def bfs(self,s):
        queue = deque()
        visited = [False]*self.vertices
        queue.append(s)
        visited[s] = True
        
        while (len(queue)):
            vertex = queue.popleft()
            for node in self.g[vertex]:
                if(not visited[node]):
                    visited[node] = True
                    queue.append(node)  
            
            print(vertex, end = ' ')
                    
            
    def dfs(self,s):
        
        stack = deque()
        visited = [False]*self.vertices
        stack.append(s)
        visited[s] = True
        
        while(len(stack)):
            vertex = stack.pop()
            for node in self.g[vertex]:
                if(not visited[node]):
                    visited[node] = True
                    stack.append(node)
                    
            print(vertex, end = ' ')  
            
    def topological_sort(self):
        stack = deque()
        result = []
        visited = [False]*self.vertices
        
        for vertex in range(self.vertices):
            if (not visited[vertex]):
                stack.append(vertex)
                visited[vertex] = True
                while(len(stack)):
                    node = stack.pop()
                    
                    for v in self.g[node]:
                        if (not visited[v]):
                            visited[v] = True
                            stack.append(v)
                            result.insert(0,v)
                result.insert(0,vertex)
        for i in result:
            print(i , end = ' ')
        
        
                
if __name__ == '__main__':
    g= Graph(6) 
    g.addEdge(5, 0); 
    g.addEdge(5, 2); 
    g.addEdge(4, 0); 
    g.addEdge(4, 1); 
    g.addEdge(2, 3); 
    g.addEdge(3, 1); 
    
    g.printGraph()
    
    '''g.bfs(5)
    print()
    g.dfs(5)
    print()'''
    g.topological_sort()